### PR TITLE
PUT Whitelist and IPv6 toggle

### PR DESCRIPTION
**Pull request by adipirro**
Added a PUT whitelist for configuring which machines are allowed to write to the cache server: #37

Also added the ability to toggle IPv6 on/off (defaulting to off). This is to avoid confusion when dealing with hybrid Ipv4-Ipv6 sockets as they convert IPv4 addresses to [Ipv4-mapped Ipv6 addresses](https://en.wikipedia.org/wiki/IPv6#IPv4-mapped_IPv6_addresses).

Relevant net library docs: https://nodejs.org/dist/latest-v4.x/docs/api/net.html#net_server_listen_port_hostname_backlog_callback

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -62,6 +62,7 @@ Command                          | Description
 `-w`, `--workers <n>`            | The number of worker threads to spawn. The default is 0.
 `-m`, `--mirror [host:port]`     | Mirror transactions to another cache server. Repeat this option ofr multiple mirrors.
 `-m`, `--monitor-parent-process <n>` | Monitor a parent process and exit if it dies.
+`--allow-ipv6`                  | Allow IPv6 connections if available.
```
Comment (stephen-palmer) [inline README.md]: I would prefer if this were in the config file - we need to move toward less discrepancy between CLI options and config file options.

To do so, I think a new top level section called "Server" would be appropriate:

```
Server
   options:
      allowIpv6: true
```

File: lib/server/server.js [hunk 2]
```diff
@@ -25,7 +25,9 @@ class CacheServer {
         if(options.mirror) {
             options.mirror = [].concat(options.mirror);
             this._mirrors = options.mirror.map(m => new TransactionMirror(m, cache));
-        }
+		}
```
Comment (stephen-palmer) [inline lib/server/server.js]: Convert tabs to spaces (there are a few other places as well)

File: lib/server/command_processor.js [hunk 3]
```diff
@@ -77,6 +82,18 @@ class CommandProcessor extends Duplex {
         Promise.resolve().then(() => this._read_internal());
     }
 
+    /**
+     * @private
+     */
+    _isWhitelisted(ip) {
+        if(this._whitelistEmpty){
```
Comment (stephen-palmer) [inline lib/server/command_processor.js]: can be simplified:

`return this._whitelistEmpty || this._putWhitelist.indexOf(ip) != -1`

File: lib/server/server.js [hunk 4]
```diff
@@ -87,7 +89,13 @@ class CacheServer {
         });
 
         return new Promise(resolve => {
-            this._server.listen(this.port, () => resolve());
+			if(this.allowIpv6){
+				this._server.listen(this.port, () => resolve());
+			}
+			else{
```
Comment (stephen-palmer) [inline lib/server/server.js]: leave a space before open braces

**Comment by adipirro** [comment]
Looks like Travis doesn't support IPv6, bummer.

I don't think there are any more tests I can reasonably add. Going to leave this as is unless you have any requests.

**Pull request merged**